# Import libraries

In [138]:
import pandas as pd
import numpy as np
from geopy.geocoders import Nominatim
import json
import requests
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

# Scraping the table from wikipedia using pandas

In [133]:
can = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0] #read the table from url

# Cleaning the data

In [134]:
can = can[can['Borough'] != 'Not assigned'] #delete rows where Borough is not assigned
can = can.reset_index() #reset index
can = can.drop(columns='index')
can.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [135]:
duplicated = can[can['Postal Code'].duplicated()] #check if there are any duplicates
duplicated.shape

(0, 3)

We see there are no duplicated rows in Postal Code column

In [137]:
can['Neighborhood'].isnull().values.any() #check if there are any empty rows in Neighborhood column

False

We see there are no empty rows in Neighborhood column.

# Printing number of rows in the dataframe

In [136]:
can.shape

(103, 3)

There are 103 distinct rows in the dataframe.

# Getting coordinates for Postal Codes

In [139]:
coords = pd.read_csv('http://cocl.us/Geospatial_data') #read csv from the link

In [142]:
can = pd.merge(can, coords, on='Postal Code') #merge dataframes
can.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494